src/solver/SparseNewtonDescentSolver.hpp

```diff
@@ -164,16 +164,6 @@ namespace cppoptlib
 				polyfem::logger().trace("linear solve residual {}", residual);
 			}
 
-			if (grad.squaredNorm() != 0 && direction.dot(grad) >= 0)
-			{
-				increase_descent_strategy();
-				polyfem::logger().log(
-					this->descent_strategy == 2 ? spdlog::level::warn : spdlog::level::debug,
-					"Newton direction is not a descent direction (Δx⋅g={}≥0); reverting to {}",
-					direction.dot(grad), this->descent_strategy_name());
-				return compute_update_direction(objFunc, x, grad, direction);
```

Comment:
I liked this being inside the newton because it could skip the extra value and grad evals and just recompute the hessian with projected PSD. But I agree it is cleaner for it to be shared in the NonlinearSolver